phoenix-core/src/main/java/org/apache/phoenix/compile/SubselectRewriter.java

```diff
@@ -202,14 +204,30 @@ private SelectStatement flatten(SelectStatement select, SelectStatement subselec
                 }
             }
         }
-        
+        OffsetNode offset = select.getOffset();
+        if (offset != null) {
+            if (offsetRewrite == null) {
+                offsetRewrite = offset;
+            } else {
+                Integer offsetValue = OffsetCompiler.compile(null, select);
+                Integer offsetValueSubselect = OffsetCompiler.compile(null, subselect);
+                if (offsetValue != null && offsetValueSubselect != null) {
+                    offsetRewrite = offsetValue < offsetValueSubselect ? offset : offsetRewrite;
+                } else {
+                    return select;
+                }
+            }
+        }
+
```

Comment:
Not sure if this would be correct. Suppose we have "select * from (select a from t offset 2 limit 8) offset 3", guess we should return the 5th row instead. If you consider the logic is too complicated to optimize at compiletime, you can simply quit flattening.